File: 0078-subsets/0078-subsets.java
class Solution {
    List<List<Integer>> ans = new ArrayList<>();
    public List<List<Integer>> subsets(int[] nums) {
        solve(nums, 0, new ArrayList<Integer>());
        return ans;
    }

    void solve(int[] arr, int index, ArrayList<Integer> curr) {
        if(arr.length == index) {
            ans.add(new ArrayList<>(curr));
            return;
        }

        //include
        curr.add(arr[index]);
        solve(arr, index + 1, curr);

        //exclude
        curr.remove(curr.size() - 1);
        solve(arr, index + 1, curr);
    }
}

File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> ans = getSubsets(nums);
        return ans;
    }

    vector<vector<int>> getSubsets(vector<int>& nums) {
        vector<vector<int>> outer;
        vector<int> t;
        outer.push_back(t);

        for (auto& num : nums) {
            int n = outer.size();
            for (int i = 0; i < n; i++) {
                vector<int> inner = outer[i];
                inner.push_back(num);
                outer.push_back(inner);
            }
        }

        return outer;
    }
};